Iterator Module Locator

mecha opened the pull request:
Attempts to incorporate a solution for #2

mecha committed:
Updated dependencies

mecha committed:
Added composer scripts for convenience

mecha committed:
Added missing `AbstractFileLocator` abstract methods

mecha committed:
Added `AbstractIteratorLocator` and test

mecha committed:
Added `AbstractRecursiveIteratorLocator` and test

mecha committed:
Added `RecursiveIteratorLocator` and test

mecha committed:
Fixed code standards

mecha committed:
Fixed use of translation traits

The `RecursiveIteratorLocator` now depends on a translator, which is used for translating strings.

XedinUnknown commented on src/AbstractRecursiveIteratorLocator.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed $configSource The config source.
```
Type is `mixed`, but arg treated as scalar.

XedinUnknown commented on src/RecursiveIteratorLocator.php:
```diff
@@ -0,0 +1,186 @@
+     * @param StringTranslatorInterface $translator The translator to use for validation messages and logging.
+     */
+    public function __construct(
+        RecursiveIterator $iterator,
```
Didn't quite understand what it iterates over. The iterator doesn't even seem to be getting set anywhere.

mecha replied:
Fixed in 44246ea3acd235422d242bcee73b7c9f1835148d

XedinUnknown replied:
@mecha, didn't understand why remove the mock. When you are testing internal protected methods, you need to instantiate the mock without calling the original constructor, and mock the other methods, i.e. getters. When you are testing the constructor, you need to instantiate the mock by calling the original constructor, and test that everything needing to be set is actually set.

XedinUnknown commented on src/RecursiveIteratorLocator.php:
```diff
@@ -0,0 +1,186 @@
+        $defaults   = $this->_getConfigDefaults();
+        $normalized = array_merge($defaults, $config);
+
+        $this->_validateConfig($normalized);
```
What do you think about doing it where the config array is being consumed, instead?

mecha replied:
Can you elaborate? Consumed by what? Do you mean at a later stage when the final module configs are returned?

XedinUnknown replied:
Consumed by anything, @mecha. There is something that consumes the result of `_normalizeConfigArray()`. That something should perhaps validate, IMHO.

XedinUnknown commented on test/functional/RecursiveRecursiveIteratorLocatorTest.php:
```diff
@@ -0,0 +1,130 @@
+ *
+ * @since [*next-version*]
+ */
+class RecursiveIteratorLocatorTest extends TestCase
```
Add test for constructor, to make sure that the things are actually getting set.

mecha committed:
Fixed filename of `RecursiveIteratorLocator` test

mecha committed:
Fixed param typehint

mecha committed:
Removed `RecursiveIteratorLocator` mock creation ...

... Now tests use the constructor to instantiate an instance of the test subject.

mecha committed:
Added constructor test

mecha committed:
Fixed translator not being set in constructor

mecha committed:
Validation of module config moved consuming code

mecha committed:
Fixed test for `RecursiveIteratorLocator`

A mock is now created and tests can choose whether or not to call the original constructor.

XedinUnknown commented:
There's no actual guarantee that config source will be an array or object. It could be a file path, or a record ID, or anything else. Therefore, [blindly JSON-encoding the source in such an abstract level](https://github.com/RebelCode/module-locator/blob/task/iterator-locator_%232/src/AbstractRecursiveIteratorLocator.php#L58) is IMHO wrong.

XedinUnknown committed:
Dependency stability increased

Tests are passing again.

XedinUnknown committed:
Test corrected

XedinUnknown commented:
@mecha, I made some changes to do 2 things:
- Tests were failing, now passing.
- Corrected the new constructor test to use getters. Please always test the getter, not the actual properties.

XedinUnknown committed:
Removed some extra slashes

XedinUnknown committed:
Fixed error in PHP 7.3.0-dev

Using `count()` for strings is not allowed.

mecha committed:
Updated dependencies

mecha committed:
Exception is thrown if config is invalid

XedinUnknown commented on src/AbstractRecursiveIteratorLocator.php:
```diff
@@ -55,7 +65,11 @@ protected function _read($config)
+            return md5(json_encode($configSource));
+        }
+
+        throw $this->_createInvalidArgumentException($this->__('Config should be an array or an object.'));
```
Should say "Config *source*".

mecha committed:
Amended exception message

XedinUnknown commented on src/AbstractRecursiveIteratorLocator.php:
```diff
@@ -0,0 +1,145 @@
+    {
+        if (!($config instanceof Iterator) && !is_array($config)) {
+            throw $this->_createModuleLocatorException(
+                $this->__('Argument must be an array or iterator.'), null
```
In a more common scenario, this would be an `InvalidArgumentException`. However, the consumer of the locator is unaware of this method and of what gets passed to it, and therefore a `ModuleLocatorExceptionInterface` is completely justified, since this is what the consumer knows may be thrown.

On the other hand, the message mentions that it's the "Argument" that is wrong. Given the above, I suggest that this is changed to read that "The config source must me an iterable". Note the word "iterable": perhaps, we can agree to use this [new PHP 7 type](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.iterable-pseudo-type) everywhere to mean e.g. `Iterator|array`.

mecha replied:
Agree on both points.

mecha committed:
Amended invalid config source exception message
